src/hnsw/utils.c

```diff
@@ -135,77 +138,91 @@ bool VersionsMatch()
+            return;
         }
 
-        query = "SELECT extversion FROM pg_extension WHERE extname = 'lantern'";
+        elog(ERROR,
+             "LanternDB binary version (%s) does not match the version in SQL (%s). This can cause errors as the "
+             "two "
+             "APIs may "
+             "differ. Please install correct binary version",
+             LDB_BINARY_VERSION,
+             catalog_version);
+    }
 
-        // Execute the query to figure out what version of lantern is in use in SQL
-        spi_result = SPI_execute(query, true, 0);
-        if(spi_result != SPI_OK_SELECT) {
-            elog(ERROR, "SPI_execute returned %s for %s", SPI_result_code_string(spi_result), query);
-        }
+    const char *query;
+    const char *version;
+    bool        isnull;
+    int         version_length;
+    int         spi_result;
+    Datum       val;
+    text       *version_text;
 
-        // Global containing the number of rows processed, should be just 1
-        if(SPI_processed != 1) {
-            elog(ERROR, "SQL version query did not return any values");
-        }
+    if(SPI_connect() != SPI_OK_CONNECT) {
+        elog(ERROR, "could not connect to executor to check binary version");
+    }
 
-        // SPI_tuptable is a global populated by SPI_execute
-        val = SPI_getbinval(SPI_tuptable->vals[ 0 ], SPI_tuptable->tupdesc, 1, &isnull);
+    query = "SELECT extversion FROM pg_extension WHERE extname = 'lantern'";
 
-        if(isnull) {
-            elog(ERROR, "Version query returned null");
-        }
+    // Execute the query to figure out what version of lantern is in use in SQL
+    spi_result = SPI_execute(query, true, 0);
+    if(spi_result != SPI_OK_SELECT) {
+        elog(ERROR, "SPI_execute returned %s for %s", SPI_result_code_string(spi_result), query);
+    }
 
-        // Grab the result and check that it matches the version in the generated header
-        version_text = DatumGetTextP(val);
-        version = text_to_cstring(version_text);
-        version_length = strlen(version);
-        if(sizeof(LDB_BINARY_VERSION) < (unsigned)version_length) {
-            version_length = sizeof(LDB_BINARY_VERSION);
-        }
+    // Global containing the number of rows processed, should be just 1
+    if(SPI_processed != 1) {
+        elog(ERROR, "SQL version query did not return any values");
+    }
 
-        comparison = strncmp(version, LDB_BINARY_VERSION, version_length);
+    // SPI_tuptable is a global populated by SPI_execute
+    val = SPI_getbinval(SPI_tuptable->vals[ 0 ], SPI_tuptable->tupdesc, 1, &isnull);
 
-        if(comparison == 0) {
-            versions_match = true;
-        }
+    if(isnull) {
+        elog(ERROR, "SQL version query returned null");
+    }
+
+    // Grab the result and check that it matches the version in the generated header
+    version_text = DatumGetTextP(val);
+    catalog_version = text_to_cstring(version_text);
+
+    if(strlen(catalog_version) == 0) {
```

Comment:
so, if catalog version has zero length, we assume it is always compatible with any binary version?
is it what we want? I would think we would consider this error in all cases.